test/index.test.js

```diff
@@ -148,6 +148,93 @@ describe('mongodb-collection-sample', function() {
     });
   });
 
+  describe('raw buffer', function() {
+    var db;
+
+    before(function(done) {
+      this.timeout(30000);
+      mongodb.MongoClient.connect('mongodb://localhost:27018/test', { useNewUrlParser: true }, function(err, client) {
+        if (err) {
+          return done(err);
+        }
+        db = client.db('test');
+
+        var docs = _range(0, 150).map(function(i) {
+          return {
+            _id: 'needle_' + i,
+            is_even: i % 2,
+            long: bson.Long.fromString('1234567890'),
+            double: 0.23456,
+            int: 1234
+          };
+        });
+        db.collection('haystack').insertMany(docs, done);
+      });
+    });
+
+    after(function(done) {
+      if (!db) {
+        return done();
+      }
+      db.dropCollection('haystack', done);
+    });
+
+    it('should return raw bson buffer when requested', function(done) {
+      sample(db, 'haystack', { size: 10, raw: true })
+        .pipe(es.through(function(doc) {
+          expect(Buffer.isBuffer(doc)).to.be.true;
+        }, function() {
+          this.emit('end');
+          done();
+        }));
+    });
+  });
+
+  // This test creates 2mil documents and samples 1mil of those to make sure
+  // buffer doesn't overflow. Only creating 2mil documents, since any more
+  // causes v8 to run out memory in heap.
+  // FYI: takes a long time to run.
+  describe('raw buffer over a large set of documents', function() {
+    var db;
+
+    before(function(done) {
+      this.timeout(300000);
+      mongodb.MongoClient.connect('mongodb://localhost:27018/test', { useNewUrlParser: true }, function(err, client) {
+        if (err) {
+          return done(err);
+        }
+        db = client.db('test');
+
+        var docs = _range(2000000).map(function(i) {
+          return {
+            _id: 'needle_' + i,
+            is_even: i % 2,
+            long: bson.Long.fromString('1234567890'),
+            double: 0.23456,
+            int: 1234
+          };
+        });
+        db.collection('haystack').insertMany(docs, done);
+      });
+    });
+
+    after(function(done) {
+      if (!db) {
+        return done();
+      }
+      db.dropCollection('haystack', done);
+    });
+
+    it('buffer does not overflow', function(done) {
+      sample(db, 'haystack', { size: 1000000, raw: true })
```

Comment:
i just decided to check if we can still sample with 1mil. seems a bit naive. is there anything specific i can check to see if buffer is overflowing?

Reply:
> 🚧 stream of conscious-too-specific answer :) 

hmm... can't think of any other than [hunting down where memory leak](https://marmelab.com/blog/2018/04/03/how-to-track-and-fix-memory-leak-with-nodejs.html) is coming from 😬 

Something that comes to mind: the actual sampling math wants `sampleSize` much lower (`5` I think $sample default i spec'd was 5)  The way the reservoir sampler is implemented here would be a big chunk of work to get rid of all the leaks. 

Even though its not a sample, the `give me 1mm documents in a bson stream without having to use something else` is neat actually. 

You could check if `opts.sampleSize > 999999` or some other flag/value on the top-level  ➡️ 
If yes, skip the server version check and return a new `BigSampler()` class that impls `BaseSampler` ➡️ 
return `db.aggregate([{$match: this.query}, {$project... etc.])` cursor wrapped in a readable stream.

Reply:
i think i lost you here 😅😅. I don't particular have a memory leak with this particular test (I don't think). And the test also running this in the `Native Sampler`. 

But you're saying I should be checking straight up if the `BaseSampler` is doing its job?